packages/validation-manager/src/ValidationManager.ts

```diff
@@ -82,128 +102,157 @@ export class ValidationManager implements IValidationManager {
+  async generateValidationResult (op: UserOperation, tracerResult: ERC7562Call): Promise<ValidationResult> {
+    // const validationData = tracerResult.calls[0].output
+    const validateUserOpCallIndex = op.factory == null ? 0 : 1
+    if (tracerResult.calls[validateUserOpCallIndex] == null) {
+      console.log('fatal: no validateuserop')
     }
 
-    const returnInfo = {
-      sigFailed: mergedValidation.aggregator !== AddressZero,
-      validUntil: mergedValidation.validUntil,
-      validAfter: mergedValidation.validAfter,
-      preOpGas: res.returnInfo.preOpGas,
-      prefund: res.returnInfo.prefund
+    // pass entrypoint and other addresses we want to resolve by name.
+    const names = {
+      ...op,
+      ep: this.entryPoint.address,
+      senderCreator: await this.entryPoint.senderCreator()
     }
-    return {
-      returnInfo,
-      senderInfo: fillEntity(userOp.sender, res.senderInfo) as StakeInfo,
-      paymasterInfo: fillEntity(userOp.paymaster, res.paymasterInfo),
-      factoryInfo: fillEntity(userOp.factory, res.factoryInfo),
-      aggregatorInfo: fillEntity(res.aggregatorInfo.aggregator, res.aggregatorInfo.stakeInfo)
+    dumpCallTree(tracerResult, names)
+
+    const validationData = this.decodeValidateUserOp(tracerResult.calls[validateUserOpCallIndex])
+    const aggregator = validationData.aggregator
+    let paymasterValidationData: ValidationData = { validAfter: 0, validUntil: maxUint48, aggregator: AddressZero }
+    if (op.paymaster != null) {
+      const validatePaymasterCallIndex = validateUserOpCallIndex + 1
+      const pmRet = this.decodeValidatePaymasterUserOp(tracerResult.calls[validatePaymasterCallIndex])
+      // paymasterContext = pmRet.context
+      paymasterValidationData = pmRet.validationData
     }
+
+    // todo: paymasterContext should be returned to parser, to validate its length.
+
+    const retStakes = await this.getStakes(op.sender, op.paymaster, op.factory)
+
+    const ret: ValidationResult = {
+      returnInfo: {
+        sigFailed: false, // can't fail here, since handleOps didn't revert.
+        validUntil: Math.min(validationData.validUntil, paymasterValidationData.validUntil),
+        validAfter: Math.max(validationData.validAfter, paymasterValidationData.validAfter),
+        preOpGas: 0, // extract from innerHandleOps parameter
+        prefund: 0 // extract from innerHandleOps parameter
+      },
+      senderInfo: retStakes.sender,
+      paymasterInfo: retStakes.paymaster,
+      factoryInfo: retStakes.factory
+    }
+    return ret
   }
 
-  // standard eth_call to simulateValidation
-  async _callSimulateValidation (userOp: UserOperation): Promise<ValidationResult> {
-    // Promise<IEntryPointSimulations.ValidationResultStructOutput> {
-    const data = entryPointSimulations.encodeFunctionData('simulateValidation', [packUserOp(userOp)])
+  async _simulateHandleOps (userOp: UserOperation): Promise<void> {
```

Comment:
Maybe don't call it "simulate" if there is still a contract called "simulation" but this contract is not used in this function?

Reply:
this is part of the code that will get away when removing unsafe.